Added file extensions property to Checker & TreeWalker, issue #25

alexkravin opened the pull request:
Added <b>fileExtensions</b> property to Checker and TreeWalker, updated sun_checks.xml

```xml
<module name="Checker">
    <property name="fileExtensions" value="java, properties, xml"/>
...
<module name="TreeWalker">

        <property name="fileExtensions" value="java"/>
...
```

Tested on pdf file:

```
$ java -jar checkstyle-6.3-SNAPSHOT-all.jar -c /sun_checks.xml Books/Spring\ in\ Action\,\ 3rd\ Edition.pdf 
Starting audit...
Audit done.
```

romani commented:
1) pliz sdelaj Utesty
2) nado obnovit` vse "xxxxxX_cheks.xml"
3) sdelaj v metodaj odnu tochku vyhoda po return

alexkravin commented:
All points are done

romani commented:
1) Coverage still decreased , please recheck why, regenerate report your self

2) 
> + if ((null == fileExtensions) || (fileExtensions.length == 0)) {
+ result = true;
+ }

"null == ...... " ? is how you pronounce in your mind ? that is practice from C++, java is not affected.
"result = true;" - as you already know result why it not the last statement in method execution for this branch ?
"if ((null == fileExtensions) || (fileExtensions.length == 0)) {" - did you described in Doc that empty property in config mean match to all ?

3) 
> private boolean fileExtensionMatches(File file)

is it copy-paste ? if yes please put that method in new or existing utils class (but not in "api" package!!!)

4) 
> + final Method fileExtensionMatches = c.getClass().getDeclaredMethod("fileExtensionMatches",
+ File.class);
+ fileExtensionMatches.setAccessible(true);

Please forget about reflection for 2-3 years. Use plain java knowledge, you have plenty of public methods in TreeWalker you should use them and do fake base on them. NO reflection till it completely impossible.

alexkravin commented:
1) Now it increased

2) I didn't change that method in the first time, now rewrote it

3) Made it static and moved to Utils.

4) Got rid of reflection, wrote UT for Checker without any problems, in FileExtensionsTest I am testing TreeWalker's reaction on fileExtensions property, but I found no way to get check side effects from method processFiltered, so in input I made 2 files with different extension (proper and not) and running this method, before running I'm checking files with fileExtensionMatches(), so all branches are covered and extensions are tested

romani commented:
>  in new or existing utils class (but not in "api" package!!!)

Please read attentively!!!, cleate Utils class at 
https://github.com/checkstyle/checkstyle/tree/master/src/main/java/com/puppycrawl/tools/checkstyle

> FileExtensionsTest.java

should be UtilsTest.java and you needto create extra file TreeWalkerTest.java that will test its usage in TreeWalker as you did changes to TreeWalker.java

> but I found no way to get check side effects from

put a fake java and expect violation event - that is side effect.

alexkravin commented:
1) Created separate Utils class in checkstyle package, moved method <b>fileExtensionMatches</b> to it

2) Created UT for CheckstyleUtils and created UT for TreeWalker, there're 2 tests with 2 fake files which have different extensions, tested for violations.

romani commented:
> CheckstyleUtils

please rename to be named as Utils.java, it is ok that name wil be similar, eventually I will move all methods from checkstyle.api.Utils to checkstyle.Utils . We need short name for util class to be easily used without static import.

 > "if ((null == fileExtensions) || (fileExtensions.length == 0)) {" - did you described in Doc that empty property in config mean match to all ?

I do not see that request from me is done.

> value=&quot;java, xml, properties&quot;/&gt;

do we have Checks that do XML validation ? I do not think so, so please remove from default value. If required user could do that himself.

alexkravin commented:
1) Ok, renamed

2) I sent you that request 2 comments ago

```java
if ((fileExtensions == null) || (fileExtensions.length == 0)) {
            result = true;
        }
        else {
            // normalize extensions so all of them have a leading dot
            final String[] withDotExtensions = new String[fileExtensions.length];
...
```

this method was rewrote.

3) Checks do xml validations, for example if there's empty line in xml which matches the pattern \s+ - Checkstyle puts violation, so I think this extension can remain.

romani commented:
> this method was rewritten

location is changed
https://github.com/alexkravin/checkstyle/commit/d4813cdecf5fe69f325933d8c9e0be293c852d56#diff-56e4e1c428966fc97ab87a8bf034a9c2R46
Please explain in xdoc what is expected behaviour for that values, do not force user to read a code.

> so I think this extension can remain.

ok.

alexkravin commented:
Xdoc updated

mkordas commented on src/main/java/com/puppycrawl/tools/checkstyle/Utils.java:
```diff
@@ -0,0 +1,72 @@
+public final class Utils
+{
+
+    /** stop instances being created **/
```
Javadocs should begin with capital letter and end with full stop.

alexkravin replied:
This was done according to https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/Utils.java#L56
